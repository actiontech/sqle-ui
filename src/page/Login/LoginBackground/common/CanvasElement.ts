import { IPosition } from './index.type';

abstract class CanvasElement {
  abstract position: IPosition;
  abstract canvas: HTMLCanvasElement;
  abstract context: CanvasRenderingContext2D;
  abstract draw(): void;
  abstract update(reachPosition: IPosition): void;
}

export default CanvasElement;
