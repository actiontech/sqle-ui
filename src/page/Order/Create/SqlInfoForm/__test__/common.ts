import instance from '../../../../../api/instance';
import { resolveThreeSecond } from '../../../../../testUtils/mockRequest';

export const mockGetInstance = () => {
  const spy = jest.spyOn(instance, 'getInstanceV2');
  spy.mockImplementation(() =>
    resolveThreeSecond({
      instance_name: 'db1',
      db_host: '20.20.20.2',
      db_port: '3306',
      db_user: 'root',
      db_type: 'mysql',
      desc: '',
      workflow_template_name: 'workflow-template-name-1',
      rule_template: {
        name: 'not_submit_test_rule33',
        is_global_rule_template: true,
      },
    })
  );

  return spy;
};
