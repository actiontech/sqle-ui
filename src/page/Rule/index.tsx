import useRequest from '@ahooksjs/use-request';
import { useTheme } from '@material-ui/styles';
import {
  Card,
  Col,
  Descriptions,
  Divider,
  PageHeader,
  Row,
  Select,
  Space,
} from 'antd';
import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';
import instance from '../../api/instance';
import ruleTemplate from '../../api/rule_template';
import EmptyBox from '../../components/EmptyBox';
import RuleList from '../../components/RuleList';
import useSyncRuleListTab from '../../components/RuleList/useSyncRuleListTab';
import useInstance from '../../hooks/useInstance';
import { Theme } from '../../types/theme.type';

const Rule = () => {
  const { updateInstanceList, generateInstanceSelectOption } = useInstance();
  const { t } = useTranslation();
  const [instanceName, setInstanceName] = useState<string | undefined>(
    undefined
  );

  const theme = useTheme<Theme>();
  const { data: instanceRules, run: getInstanceRules } = useRequest(
    () =>
      instance.getInstanceRuleListV1({
        instance_name: instanceName ?? '',
      }),
    {
      manual: true,
      formatResult(res) {
        return res.data?.data ?? [];
      },
    }
  );

  const { data: allRules } = useRequest(
    ruleTemplate.getRuleListV1.bind(ruleTemplate),
    {
      formatResult(res) {
        return res.data?.data ?? [];
      },
    }
  );

  const disableRules = React.useMemo(() => {
    if (!instanceRules) {
      return allRules ?? [];
    }
    const all = allRules ?? [];
    return all.filter(
      (e) => !instanceRules.find((item) => item.rule_name === e.rule_name)
    );
  }, [allRules, instanceRules]);

  React.useEffect(() => {
    if (instanceName !== undefined) {
      getInstanceRules();
    }
  }, [getInstanceRules, instanceName]);

  React.useEffect(() => {
    updateInstanceList();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const { tabKey, allTypes, tabChange } = useSyncRuleListTab(allRules);

  return (
    <>
      <PageHeader title={t('rule.pageTitle')} ghost={false}>
        {t('rule.pageDesc')}
      </PageHeader>
      <section className="padding-content">
        <Space
          size={theme.common.padding}
          direction="vertical"
          className="full-width-element"
        >
          <Card>
            <Row align="middle">
              <Col span={3}>{t('rule.form.instance')}</Col>
              <Col span={5}>
                <Select
                  data-testid="instance-name"
                  value={instanceName}
                  onChange={setInstanceName}
                  placeholder={t('common.form.placeholder.select')}
                  className="full-width-element"
                  allowClear
                >
                  {generateInstanceSelectOption()}
                </Select>
              </Col>
            </Row>
          </Card>
          <EmptyBox if={!instanceName}>
            <Card title={t('rule.allRules')}>
              <RuleList list={allRules ?? []} />
            </Card>
          </EmptyBox>
          <EmptyBox if={!!instanceName}>
            <Card title={t('rule.instanceRuleList')}>
              <Descriptions
                title={t('rule.activeRules', { name: instanceName })}
              />
              <RuleList
                list={instanceRules ?? []}
                allRuleTabs={allTypes}
                currentTab={tabKey}
                tabChange={tabChange}
              />
              <Divider dashed />
              <Descriptions
                title={t('rule.disableRules', { name: instanceName })}
              />
              <RuleList
                list={disableRules ?? []}
                allRuleTabs={allTypes}
                currentTab={tabKey}
                tabChange={tabChange}
              />
            </Card>
          </EmptyBox>
        </Space>
      </section>
    </>
  );
};

export default Rule;
