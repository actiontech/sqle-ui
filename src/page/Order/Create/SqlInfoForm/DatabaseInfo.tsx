import { MinusCircleOutlined, PlusOutlined } from '@ant-design/icons';
import { Button, Col, Form, Row, Select } from 'antd';
import { cloneDeep } from 'lodash';
import { useEffect, useRef, useState } from 'react';
import { useTranslation } from 'react-i18next';
import instance from '../../../../api/instance';
import { getInstanceTipListV1FunctionalModuleEnum } from '../../../../api/instance/index.enum';
import EmptyBox from '../../../../components/EmptyBox';
import { ResponseCode } from '../../../../data/common';
import useInstance from '../../../../hooks/useInstance';
import { DatabaseInfoProps, SchemaListType } from './index.type';

const DatabaseInfo: React.FC<DatabaseInfoProps> = ({
  form,
  instanceNameChange,
  setInstanceNames,
  setChangeSqlModeDisabled,
  clearTaskInfoWithKey,
  projectName,
}) => {
  const { t } = useTranslation();

  const [schemaList, setSchemaList] = useState<SchemaListType>(
    new Map([[0, []]])
  );

  const instanceTypeMap = useRef<Map<number, string>>(new Map());
  const { updateInstanceList, generateInstanceSelectOption, instanceList } =
    useInstance();

  const handleInstanceNameChange = (name: string, index: number) => {
    setInstanceNames((values) => {
      const cloneValue = cloneDeep(values);
      cloneValue.set(index, name);
      return cloneValue;
    });

    instanceNameChange?.(name);
    updateSchemaList(name, index);
    const currentInstance = instanceList.find((v) => v.instance_name === name);

    getInstanceTypeWithAction(index, 'add', currentInstance?.instance_type);
  };

  const updateSchemaList = (name: string, index: number) => {
    instance
      .getInstanceSchemasV1({
        instance_name: name,
        project_name: projectName,
      })
      .then((res) => {
        if (res.data.code === ResponseCode.SUCCESS) {
          setSchemaList((values) => {
            const cloneValue = cloneDeep(values);
            cloneValue.set(index, res.data.data?.schema_name_list ?? []);
            return cloneValue;
          });
        }
      });
  };

  const generateInstanceSchemaSelectOption = (index: number) => {
    return (
      schemaList.get(index)?.map((schema) => (
        <Select.Option value={schema} key={schema}>
          {schema}
        </Select.Option>
      )) ?? []
    );
  };

  const getInstanceTypeWithAction = (
    key: number,
    type: 'add' | 'remove',
    instanceType = ''
  ) => {
    if (type === 'add') {
      instanceTypeMap.current.set(key, instanceType);
    } else if (type === 'remove') {
      instanceTypeMap.current.delete(key);
    }
    const instanceTypeSet = new Set(instanceTypeMap.current.values());
    const isExistDifferentInstanceType = instanceTypeSet.size > 1;
    setChangeSqlModeDisabled(isExistDifferentInstanceType);
  };

  useEffect(() => {
    updateInstanceList({
      project_name: projectName,
      functional_module:
        getInstanceTipListV1FunctionalModuleEnum.create_workflow,
    });
  }, [projectName, updateInstanceList]);

  return (
    <Form.List name="dataBaseInfo" initialValue={[{}]}>
      {(fields, { add, remove }) => (
        <>
          {fields.map((field, index) => (
            <Row key={field.key}>
              <Col span={12}>
                <Form.Item
                  labelCol={{
                    xs: { span: 24 },
                    sm: { span: 16 },
                  }}
                  wrapperCol={{
                    xs: { span: 24 },
                    sm: { span: 8 },
                  }}
                  label={t('order.sqlInfo.instanceName')}
                  {...field}
                  name={[field.name, 'instanceName']}
                  rules={[
                    {
                      required: true,
                    },
                  ]}
                >
                  <Select<string>
                    onChange={(value) =>
                      handleInstanceNameChange(value, field.key)
                    }
                    showSearch={true}
                    placeholder={t('common.form.placeholder.select', {
                      name: t('order.sqlInfo.instanceName'),
                    })}
                  >
                    {generateInstanceSelectOption()}
                  </Select>
                </Form.Item>
              </Col>
              <Col span={6}>
                <Row gutter={8}>
                  <Col span={22}>
                    <Form.Item
                      noStyle
                      shouldUpdate={(prevValues, curValues) => {
                        return (
                          prevValues?.dataBaseInfo[index]?.instanceName !==
                          curValues?.dataBaseInfo[index]?.instanceName
                        );
                      }}
                    >
                      {() => (
                        <Form.Item
                          labelCol={{
                            xs: { span: 24 },
                            sm: { span: 8 },
                          }}
                          wrapperCol={{
                            xs: { span: 24 },
                            sm: { span: 24 },
                            md: { span: 24 },
                          }}
                          {...field}
                          name={[field.name, 'instanceSchema']}
                          label={t('order.sqlInfo.instanceSchema')}
                        >
                          <Select
                            disabled={
                              !form.getFieldValue('dataBaseInfo')[index]
                                ?.instanceName
                            }
                            placeholder={t('common.form.placeholder.select')}
                            showSearch
                            allowClear
                          >
                            {generateInstanceSchemaSelectOption(index)}
                          </Select>
                        </Form.Item>
                      )}
                    </Form.Item>
                  </Col>
                  <Col span={2}>
                    <EmptyBox if={index !== 0}>
                      <MinusCircleOutlined
                        style={{ marginTop: 8 }}
                        data-testid="remove-item"
                        onClick={() => {
                          setInstanceNames((values) => {
                            const cloneValue = cloneDeep(values);
                            cloneValue.delete(field.key);
                            return cloneValue;
                          });
                          getInstanceTypeWithAction(field.key, 'remove');
                          remove(index);
                          clearTaskInfoWithKey(field.key.toString());
                        }}
                      />
                    </EmptyBox>
                  </Col>
                </Row>
              </Col>
            </Row>
          ))}
          {/* IFTRUE_isEE */}
          <Form.Item label=" " colon={false}>
            <Button
              type="dashed"
              onClick={() => {
                setInstanceNames((values) => {
                  const cloneValue = cloneDeep(values);
                  cloneValue.set(fields[fields.length - 1].key + 1, '');
                  return cloneValue;
                });
                add();
              }}
              block
              icon={<PlusOutlined />}
            >
              {t('order.sqlInfo.addInstance')}
            </Button>
          </Form.Item>
          {/* FITRUE_isEE */}
        </>
      )}
    </Form.List>
  );
};

export default DatabaseInfo;
