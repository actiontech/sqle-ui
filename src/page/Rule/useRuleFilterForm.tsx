import { Select } from 'antd';
import { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useLocation } from 'react-router-dom';
import { IRuleResV1 } from '../../api/common';
import useDatabaseType from '../../hooks/useDatabaseType';
import useGlobalRuleTemplate from '../../hooks/useGlobalRuleTemplate';
import useProject from '../../hooks/useProject';
import useRuleTemplate from '../../hooks/useRuleTemplate';

export enum RuleUrlParamKey {
  projectName = 'projectName',
  ruleTemplateName = 'ruleTemplateName',
}

const useRuleFilterForm = (
  getProjectTemplateRules: (
    projectName?: string,
    ruleTemplateName?: string
  ) => Promise<IRuleResV1[]>,
  getGlobalTemplateRules: (ruleTemplateName?: string) => Promise<IRuleResV1[]>
) => {
  const { t } = useTranslation();
  const location = useLocation();

  const [dbType, setDbType] = useState<string | undefined>(undefined);
  const { updateDriverNameList, driverNameList, generateDriverSelectOptions } =
    useDatabaseType();
  const { updateProjectList, generateProjectSelectOption } = useProject();
  const { ruleTemplateList, updateRuleTemplateList } = useRuleTemplate();
  const { globalRuleTemplateList, updateGlobalRuleTemplateList } =
    useGlobalRuleTemplate();

  const [projectName, setProjectName] = useState<string>();
  const [ruleTemplateName, setRuleTemplateName] = useState<string>();

  const projectNameChangeHandle = (name: string) => {
    setProjectName(name);
    setRuleTemplateName(undefined);
    if (name) {
      updateRuleTemplateList(name);
    }
  };

  const ruleTemplateNameChangeHandle = (name: string) => {
    setRuleTemplateName(name);
    if (!name) {
      return;
    }
    if (projectName) {
      getProjectTemplateRules(projectName, name);
    } else {
      getGlobalTemplateRules(name);
    }
  };

  const generateRuleTemplateSelectOptions = () => {
    const list = projectName ? ruleTemplateList : globalRuleTemplateList;
    const groupLabel = projectName
      ? t('rule.projectRuleTemplate')
      : t('rule.globalRuleTemplate');

    if (list.length === 0) {
      return null;
    }

    return (
      <Select.OptGroup label={groupLabel}>
        {list.map((v) => {
          return (
            <Select.Option
              key={v.rule_template_name}
              value={v.rule_template_name ?? ''}
            >
              {v.rule_template_name}
            </Select.Option>
          );
        })}
      </Select.OptGroup>
    );
  };

  useEffect(() => {
    updateGlobalRuleTemplateList();
    updateDriverNameList();
    updateProjectList();
  }, [updateDriverNameList, updateGlobalRuleTemplateList, updateProjectList]);

  useEffect(() => {
    if (driverNameList.length > 0 && !dbType) {
      setDbType(driverNameList[0]);
    }
  }, [driverNameList, dbType]);

  useEffect(() => {
    const searchParams = new URLSearchParams(location.search);

    const projectNameInUrl = searchParams.get(RuleUrlParamKey.projectName);
    const ruleTemplateNameInUrl = searchParams.get(
      RuleUrlParamKey.ruleTemplateName
    );
    if (projectNameInUrl) {
      setProjectName(projectNameInUrl);
      updateRuleTemplateList(projectNameInUrl);
    }

    if (ruleTemplateNameInUrl) {
      setRuleTemplateName(ruleTemplateNameInUrl);
      projectNameInUrl
        ? getProjectTemplateRules(projectNameInUrl, ruleTemplateNameInUrl)
        : getGlobalTemplateRules(ruleTemplateNameInUrl);
    }
  }, [
    getGlobalTemplateRules,
    getProjectTemplateRules,
    location.search,
    updateRuleTemplateList,
  ]);

  return {
    generateDriverSelectOptions,
    generateProjectSelectOption,
    generateRuleTemplateSelectOptions,
    projectName,
    projectNameChangeHandle,
    dbType,
    setDbType,
    ruleTemplateName,
    ruleTemplateNameChangeHandle,
  };
};

export default useRuleFilterForm;
