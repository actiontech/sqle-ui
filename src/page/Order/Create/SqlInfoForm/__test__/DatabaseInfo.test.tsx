import { act, fireEvent, screen } from '@testing-library/react';
import Form, { useForm } from 'antd/lib/form/Form';
import { getInstanceTipListV1FunctionalModuleEnum } from '../../../../../api/instance/index.enum';
import { renderWithRouter } from '../../../../../testUtils/customRender';
import {
  mockInstanceTip,
  mockUseInstanceSchema,
  resolveThreeSecond,
} from '../../../../../testUtils/mockRequest';
import { RuleUrlParamKey } from '../../../../Rule/useRuleFilterForm';
import DatabaseInfo from '../DatabaseInfo';
import { renderHook } from '@testing-library/react-hooks';
import { mockGetInstance } from './common';

const selectOptionByIndex = (
  label: string,
  optionText: string,
  index = 1,
  labelIndex = 0
) => {
  fireEvent.mouseDown(screen.getAllByLabelText(label)[labelIndex]);
  const options = screen.getAllByText(optionText);
  let realIndex = index;
  if (index < 0) {
    realIndex = options.length + index;
  }
  const option = options[realIndex];
  expect(option).toHaveClass('ant-select-item-option-content');
  fireEvent.click(option);
};

describe('test Order/Create/SqlInfoForm/DatabaseInfo', () => {
  const instanceNameChange = jest.fn();
  const setInstanceNames = jest.fn();
  const setChangeSqlModeDisabled = jest.fn();
  const clearTaskInfoWithKey = jest.fn();
  const projectName = 'default';
  let getInstanceListSpy: jest.SpyInstance;
  let getInstanceSchemaSpy: jest.SpyInstance;
  let getInstanceSpy: jest.SpyInstance;

  beforeEach(() => {
    jest.useFakeTimers();
    getInstanceListSpy = mockInstanceTip();
    getInstanceSchemaSpy = mockUseInstanceSchema();
    getInstanceSpy = mockGetInstance();
  });

  afterEach(() => {
    jest.useRealTimers();
    jest.clearAllMocks();
    jest.clearAllTimers();
  });

  const renderComponent = () => {
    const { result } = renderHook(() => useForm());
    return renderWithRouter(
      <Form form={result.current[0]}>
        <DatabaseInfo
          form={result.current[0]}
          instanceNameChange={instanceNameChange}
          setChangeSqlModeDisabled={setChangeSqlModeDisabled}
          setInstanceNames={setInstanceNames}
          clearTaskInfoWithKey={clearTaskInfoWithKey}
          projectName={projectName}
        />
      </Form>
    );
  };

  test('should match snapshot', async () => {
    const { container } = renderComponent();

    await act(async () => jest.advanceTimersByTime(3000));

    expect(container).toMatchSnapshot();
  });

  test('should be able to add and remove items', async () => {
    expect(getInstanceListSpy).toBeCalledTimes(0);
    expect(getInstanceSchemaSpy).toBeCalledTimes(0);

    renderComponent();

    expect(instanceNameChange).toBeCalledTimes(0);
    expect(setChangeSqlModeDisabled).toBeCalledTimes(0);
    expect(setInstanceNames).toBeCalledTimes(0);
    expect(clearTaskInfoWithKey).toBeCalledTimes(0);

    expect(getInstanceSpy).toBeCalledTimes(0);
    expect(getInstanceListSpy).toBeCalledTimes(1);
    expect(getInstanceListSpy).toBeCalledWith({
      project_name: projectName,
      functional_module:
        getInstanceTipListV1FunctionalModuleEnum.create_workflow,
    });
    expect(getInstanceSchemaSpy).toBeCalledTimes(0);
    await act(async () => jest.advanceTimersByTime(3000));

    expect(
      screen.queryAllByLabelText('order.sqlInfo.instanceName').length
    ).toBe(1);
    expect(
      screen.queryAllByLabelText('order.sqlInfo.instanceSchema').length
    ).toBe(1);
    expect(
      screen.queryAllByLabelText('order.sqlInfo.instanceSchema')[0]
    ).toBeDisabled();
    expect(screen.queryByTestId('remove-item')).not.toBeInTheDocument();

    selectOptionByIndex('order.sqlInfo.instanceName', 'mysql-test');
    await act(async () => jest.advanceTimersByTime(0));

    expect(setInstanceNames).toBeCalledTimes(1);
    expect(setInstanceNames.mock.calls[0][0](new Map([[0, '']]))).toEqual(
      new Map([[0, 'mysql-test']])
    );

    expect(instanceNameChange).toBeCalledTimes(1);
    expect(instanceNameChange).toBeCalledWith('mysql-test');

    expect(setChangeSqlModeDisabled).toBeCalledTimes(1);
    expect(setChangeSqlModeDisabled).toBeCalledWith(false);

    expect(getInstanceSchemaSpy).toBeCalledTimes(1);
    expect(getInstanceSchemaSpy).toBeCalledWith({
      instance_name: 'mysql-test',
      project_name: projectName,
    });

    expect(getInstanceSpy).toBeCalledTimes(1);
    expect(getInstanceSpy).toBeCalledWith({
      instance_name: 'mysql-test',
      project_name: projectName,
    });

    expect(
      screen.queryAllByLabelText('order.sqlInfo.instanceSchema')[0]
    ).not.toBeDisabled();

    await act(async () => jest.advanceTimersByTime(3000));

    expect(
      screen.getByText('rule.form.ruleTemplate: not_submit_test_rule33')
    ).toBeInTheDocument();

    expect(
      screen
        .getByText('rule.form.ruleTemplate: not_submit_test_rule33')
        .getAttribute('href')
    ).toBe(`/rule?${RuleUrlParamKey.ruleTemplateName}=not_submit_test_rule33`);

    selectOptionByIndex('order.sqlInfo.instanceSchema', 'schema1');
    await act(async () => jest.advanceTimersByTime(0));

    //add
    fireEvent.click(screen.getByText('order.sqlInfo.addInstance'));

    expect(setInstanceNames).toBeCalledTimes(2);
    expect(
      setInstanceNames.mock.calls[1][0](new Map([[0, 'mysql-test']]))
    ).toEqual(
      new Map([
        [0, 'mysql-test'],
        [1, ''],
      ])
    );

    expect(
      screen.queryAllByLabelText('order.sqlInfo.instanceName').length
    ).toBe(2);
    expect(
      screen.queryAllByLabelText('order.sqlInfo.instanceSchema').length
    ).toBe(2);
    expect(screen.getByTestId('remove-item')).toBeInTheDocument();

    selectOptionByIndex('order.sqlInfo.instanceName', 'mysql-test', 4, 1);
    await act(async () => jest.advanceTimersByTime(0));

    await act(async () => jest.advanceTimersByTime(3000));

    expect(setInstanceNames).toBeCalledTimes(3);
    expect(
      setInstanceNames.mock.calls[2][0](
        new Map([
          [0, 'mysql-test'],
          [1, ''],
        ])
      )
    ).toEqual(
      new Map([
        [0, 'mysql-test'],
        [1, 'mysql-test'],
      ])
    );

    expect(instanceNameChange).toBeCalledTimes(2);

    expect(setChangeSqlModeDisabled).toBeCalledTimes(2);
    expect(setChangeSqlModeDisabled).nthCalledWith(2, false);

    expect(getInstanceSchemaSpy).toBeCalledTimes(2);
    expect(getInstanceSchemaSpy).nthCalledWith(2, {
      instance_name: 'mysql-test',
      project_name: projectName,
    });

    expect(getInstanceSpy).toBeCalledTimes(2);
    expect(getInstanceSpy).nthCalledWith(2, {
      instance_name: 'mysql-test',
      project_name: projectName,
    });
    expect(
      screen.getAllByText('rule.form.ruleTemplate: not_submit_test_rule33')
        .length
    ).toBe(2);

    //remove
    fireEvent.click(screen.getByTestId('remove-item'));

    expect(
      screen.queryAllByLabelText('order.sqlInfo.instanceName').length
    ).toBe(1);
    expect(
      screen.queryAllByLabelText('order.sqlInfo.instanceSchema').length
    ).toBe(1);
    expect(
      screen.getAllByText('rule.form.ruleTemplate: not_submit_test_rule33')
        .length
    ).toBe(1);

    expect(screen.queryByTestId('remove-item')).not.toBeInTheDocument();

    expect(setInstanceNames).toBeCalledTimes(4);
    expect(
      setInstanceNames.mock.calls[3][0](
        new Map([
          [0, 'mysql-test'],
          [1, 'mysql-test'],
        ])
      )
    ).toEqual(new Map([[0, 'mysql-test']]));

    expect(setChangeSqlModeDisabled).toBeCalledTimes(3);
    expect(setChangeSqlModeDisabled).toBeCalledWith(false);

    expect(clearTaskInfoWithKey).toBeCalledTimes(1);
    expect(clearTaskInfoWithKey).toBeCalledWith('1');

    //add
    fireEvent.click(screen.getByText('order.sqlInfo.addInstance'));

    expect(setInstanceNames).toBeCalledTimes(5);
    expect(
      setInstanceNames.mock.calls[4][0](new Map([[0, 'mysql-test']]))
    ).toEqual(
      new Map([
        [0, 'mysql-test'],
        [1, ''],
      ])
    );

    selectOptionByIndex('order.sqlInfo.instanceName', 'oracle-test', 1, 1);
    await act(async () => jest.advanceTimersByTime(0));

    await act(async () => jest.advanceTimersByTime(3000));

    expect(setInstanceNames).toBeCalledTimes(6);
    expect(
      setInstanceNames.mock.calls[5][0](
        new Map([
          [0, 'mysql-test'],
          [1, ''],
        ])
      )
    ).toEqual(
      new Map([
        [0, 'mysql-test'],
        [1, ''],
        [2, 'oracle-test'],
      ])
    );

    expect(instanceNameChange).toBeCalledTimes(3);

    expect(setChangeSqlModeDisabled).toBeCalledTimes(4);
    expect(setChangeSqlModeDisabled).nthCalledWith(4, true);

    expect(getInstanceSchemaSpy).toBeCalledTimes(3);
    expect(getInstanceSchemaSpy).nthCalledWith(3, {
      instance_name: 'oracle-test',
      project_name: projectName,
    });

    expect(getInstanceSpy).toBeCalledTimes(3);
    expect(getInstanceSpy).nthCalledWith(3, {
      instance_name: 'oracle-test',
      project_name: projectName,
    });

    //remove
    fireEvent.click(screen.getByTestId('remove-item'));

    expect(setInstanceNames).toBeCalledTimes(7);
    expect(
      setInstanceNames.mock.calls[6][0](
        new Map([
          [0, 'mysql-test'],
          [1, ''],
          [2, 'oracle-test'],
        ])
      )
    ).toEqual(
      new Map([
        [0, 'mysql-test'],
        [1, ''],
      ])
    );

    expect(setChangeSqlModeDisabled).toBeCalledTimes(5);
    expect(setChangeSqlModeDisabled).toBeCalledWith(false);

    expect(clearTaskInfoWithKey).toBeCalledTimes(2);
    expect(clearTaskInfoWithKey).toBeCalledWith('2');
  });

  test('should set project name when rule template is not global', async () => {
    getInstanceSpy.mockImplementation(() =>
      resolveThreeSecond({
        rule_template: {
          is_global_rule_template: false,
          name: 'default_rule_template',
        },
      })
    );

    renderComponent();

    await act(async () => jest.advanceTimersByTime(3000));

    selectOptionByIndex('order.sqlInfo.instanceName', 'mysql-test');
    await act(async () => jest.advanceTimersByTime(3000));

    expect(getInstanceSpy).toBeCalled();
    await act(async () => jest.advanceTimersByTime(3000));

    expect(
      screen.getByText('rule.form.ruleTemplate: default_rule_template')
    ).toBeInTheDocument();
    expect(
      screen
        .getByText('rule.form.ruleTemplate: default_rule_template')
        .getAttribute('href')
    ).toBe(
      `/rule?${RuleUrlParamKey.ruleTemplateName}=default_rule_template&${RuleUrlParamKey.projectName}=${projectName}`
    );
  });
});
