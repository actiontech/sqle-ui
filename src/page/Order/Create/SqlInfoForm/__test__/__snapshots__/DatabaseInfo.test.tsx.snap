// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Order/Create/SqlInfoForm/DatabaseInfo should match snapshot 1`] = `
<div>
  <form
    class="ant-form ant-form-horizontal"
  >
    <div
      class="ant-row"
    >
      <div
        class="ant-col ant-col-12"
      >
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label ant-col-xs-24 ant-col-sm-16"
          >
            <label
              class="ant-form-item-required"
              for="dataBaseInfo_0_instanceName"
              title="order.sqlInfo.instanceName"
            >
              order.sqlInfo.instanceName
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control ant-col-xs-24 ant-col-sm-8"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-select ant-select-single ant-select-show-arrow ant-select-show-search"
                >
                  <div
                    class="ant-select-selector"
                  >
                    <span
                      class="ant-select-selection-search"
                    >
                      <input
                        aria-activedescendant="dataBaseInfo_0_instanceName_list_0"
                        aria-autocomplete="list"
                        aria-controls="dataBaseInfo_0_instanceName_list"
                        aria-haspopup="listbox"
                        aria-owns="dataBaseInfo_0_instanceName_list"
                        autocomplete="off"
                        class="ant-select-selection-search-input"
                        id="dataBaseInfo_0_instanceName"
                        role="combobox"
                        type="search"
                        value=""
                      />
                    </span>
                    <span
                      class="ant-select-selection-placeholder"
                    >
                      common.form.placeholder.select
                    </span>
                  </div>
                  <span
                    aria-hidden="true"
                    class="ant-select-arrow"
                    style="user-select: none;"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down ant-select-suffix"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-col ant-col-6"
      >
        <div
          class="ant-row"
          style="margin-left: -4px; margin-right: -4px;"
        >
          <div
            class="ant-col ant-col-22"
            style="padding-left: 4px; padding-right: 4px;"
          >
            <div
              class="ant-row ant-form-item"
            >
              <div
                class="ant-col ant-form-item-label ant-col-xs-24 ant-col-sm-8"
              >
                <label
                  class=""
                  for="dataBaseInfo_0_instanceSchema"
                  title="order.sqlInfo.instanceSchema"
                >
                  order.sqlInfo.instanceSchema
                </label>
              </div>
              <div
                class="ant-col ant-form-item-control ant-col-xs-24 ant-col-sm-24 ant-col-md-24"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <div
                      class="ant-select ant-select-single ant-select-allow-clear ant-select-show-arrow ant-select-disabled ant-select-show-search"
                    >
                      <div
                        class="ant-select-selector"
                      >
                        <span
                          class="ant-select-selection-search"
                        >
                          <input
                            aria-activedescendant="dataBaseInfo_0_instanceSchema_list_0"
                            aria-autocomplete="list"
                            aria-controls="dataBaseInfo_0_instanceSchema_list"
                            aria-expanded="false"
                            aria-haspopup="listbox"
                            aria-owns="dataBaseInfo_0_instanceSchema_list"
                            autocomplete="off"
                            class="ant-select-selection-search-input"
                            disabled=""
                            id="dataBaseInfo_0_instanceSchema"
                            role="combobox"
                            type="search"
                            value=""
                          />
                        </span>
                        <span
                          class="ant-select-selection-placeholder"
                        >
                          common.form.placeholder.select
                        </span>
                      </div>
                      <span
                        aria-hidden="true"
                        class="ant-select-arrow"
                        style="user-select: none;"
                        unselectable="on"
                      >
                        <span
                          aria-label="down"
                          class="anticon anticon-down ant-select-suffix"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="down"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                            />
                          </svg>
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-col ant-col-2"
            style="padding-left: 4px; padding-right: 4px;"
          />
        </div>
      </div>
    </div>
    <div
      class="ant-row ant-form-item"
    >
      <div
        class="ant-col ant-form-item-label"
      >
        <label
          class="ant-form-item-no-colon"
          title=" "
        >
           
        </label>
      </div>
      <div
        class="ant-col ant-form-item-control"
      >
        <div
          class="ant-form-item-control-input"
        >
          <div
            class="ant-form-item-control-input-content"
          >
            <button
              class="ant-btn ant-btn-dashed ant-btn-block"
              type="button"
            >
              <span
                aria-label="plus"
                class="anticon anticon-plus"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="plus"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <defs>
                    <style />
                  </defs>
                  <path
                    d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
                  />
                  <path
                    d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
                  />
                </svg>
              </span>
              <span>
                order.sqlInfo.addInstance
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
